### PR TITLE
Custom Attention Layer + Fix some stuff

### PR DESCRIPTION
**Pull request by Gurmie12**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -48,4 +49,30 @@ def forward(self, x, pad_mask):
         x = self.output_norm(x)
         x = self.output(x)
         return torch.squeeze(x)
-    
\ No newline at end of file
+
+class WatsonCrickAttentionLayer(torch.nn.Module):
+
+    def __init__(self, size, score_matrix):
```
Comment (bl21812) [inline models.py:55]: not sure why score_matrix needs to be passed in? this should probably just be randomly initialized like any other params

File: dataset.py [hunk 2]
```diff
@@ -80,3 +82,75 @@ def __getitem__(self, idx):
             label = label.to(self.device)
 
         return inp, label, pad_mask
+
+
+class BPPInputDataset(Dataset):
+
```
Comment (bl21812) [inline dataset.py:88]: given that this approach needs bpp and sequence data, the dataset should store and return sequence, bpp, and label (basically it should be the RNAInputDataset above but with the associated BPP matrix for each sequence as well)

File: models.py [hunk 3]
```diff
@@ -48,4 +49,30 @@ def forward(self, x, pad_mask):
...
+        super(WatsonCrickAttentionLayer, self).__init__()
+        self.size = size
+        self.score_matrix = torch.nn.Parameter(score_matrix, requires_grad=False)
+
+        # Linear Components
+        self.q_linear = torch.nn.Linear(size, size)
+        self.k_linear = torch.nn.Linear(size, size)
+        self.v_linear = torch.nn.Linear(size, size)
+
+    def forward(self, x):
```
Comment (bl21812) [inline models.py:65]: which of q, k, v is supposed to be the sequence / bpp matrix?

**Review by bl21812 — CHANGES_REQUESTED** [review]

File: models.py (hunk 3 above)
Comment (Gurmie12) [reply models.py:65]: it could be query or key I believe

File: dataset.py (hunk 2 above)
Comment (Gurmie12) [reply dataset.py:88]: This was from Finn's branch during my re-base so might be worth asking him? I could also alternatively re-write it if he's ok with that
